## Установка нужных версий библиотек

In [33]:
!wget 'https://drive.google.com/uc?export=download&id=1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn' -O requirements.txt

--2026-07-29 23:42:52--  https://drive.google.com/uc?export=download&id=1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn
Resolving drive.google.com (drive.google.com)... 192.178.210.101, 192.178.210.139, 192.178.210.113, ...
Connecting to drive.google.com (drive.google.com)|192.178.210.101|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn&export=download [following]
--2026-07-29 23:42:52--  https://drive.usercontent.google.com/download?id=1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn&export=download
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 192.178.210.132, 2607:f8b0:4001:c6e::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|192.178.210.132|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 74 [application/octet-stream]
Saving to: ‘requirements.txt’

requirements.txt    100%[===================>]      74  --.-KB/s    

In [34]:
!pip install -r requirements.txt

  Using cached scikit-learn-0.22.2.post1.tar.gz (6.9 MB)
  Preparing metadata (setup.py) ... done
  Using cached pandas-1.1.5.tar.gz (5.2 MB)
  Installing build dependencies ... done
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.


In [35]:
import pandas as pd
import numpy as np

Метрики в машинном обучении нужны для того, чтобы оценивать качество работы модели, чтобы понимать, а насколько модель хорошо запомнила закономерности из обучающей выборки.  

Сегодня будем рассматривать метрики для задачи регрессии - предсказания непрерывной целевой переменной. Для такой задачи хочется, чтобы наши модели предсказывали числа как можно ближе к истинным, чтобы отклонения реального значения и прогноза модели были минимальными.

<img src='https://miro.medium.com/max/864/0*OVG0UBquPwSiwiUi' width=600>


Для подсчета разных метрик регрессии возьмем данные из стандартных датасетов в `sklearn`.

## Получение данных

Будем работать с набором данных для задачи регрессии (целевая переменная - стоимость дома) `load_boston`, который можно получить из стандартных датасетов в `sklearn'e`.

После `load_boston()` возвращается словарь с данными (`data`), целевой переменной (`target`), названиями характеристик в данных (`feature_names`) и описанием данных (`DESCR`).

In [36]:
from sklearn.datasets import load_boston

data = load_boston()
data.keys()

ImportError: 
`load_boston` has been removed from scikit-learn since version 1.2.

The Boston housing prices dataset has an ethical problem: as
investigated in [1], the authors of this dataset engineered a
non-invertible variable "B" assuming that racial self-segregation had a
positive impact on house prices [2]. Furthermore the goal of the
research that led to the creation of this dataset was to study the
impact of air quality but it did not give adequate demonstration of the
validity of this assumption.

The scikit-learn maintainers therefore strongly discourage the use of
this dataset unless the purpose of the code is to study and educate
about ethical issues in data science and machine learning.

In this special case, you can fetch the dataset from the original
source::

    import pandas as pd
    import numpy as np

    data_url = "http://lib.stat.cmu.edu/datasets/boston"
    raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
    data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
    target = raw_df.values[1::2, 2]

Alternative datasets include the California housing dataset and the
Ames housing dataset. You can load the datasets as follows::

    from sklearn.datasets import fetch_california_housing
    housing = fetch_california_housing()

for the California housing dataset and::

    from sklearn.datasets import fetch_openml
    housing = fetch_openml(name="house_prices", as_frame=True)

for the Ames housing dataset.

[1] M Carlisle.
"Racist data destruction?"
<https://medium.com/@docintangible/racist-data-destruction-113e3eff54a8>

[2] Harrison Jr, David, and Daniel L. Rubinfeld.
"Hedonic housing prices and the demand for clean air."
Journal of environmental economics and management 5.1 (1978): 81-102.
<https://www.researchgate.net/publication/4974606_Hedonic_housing_prices_and_the_demand_for_clean_air>


In [46]:
from sklearn.datasets import fetch_openml

# Сохраняем весь скачанный объект Bunch сразу в переменную data
data = fetch_openml(name='boston', version=1, as_frame=True)

In [48]:
# Вывод описания
print(data.DESCR)

**Author**:   
**Source**: Unknown - Date unknown  
**Please cite**:   

The Boston house-price data of Harrison, D. and Rubinfeld, D.L. 'Hedonic
prices and the demand for clean air', J. Environ. Economics & Management,
vol.5, 81-102, 1978.   Used in Belsley, Kuh & Welsch, 'Regression diagnostics
...', Wiley, 1980.   N.B. Various transformations are used in the table on
pages 244-261 of the latter.
Variables in order:
CRIM     per capita crime rate by town
ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
INDUS    proportion of non-retail business acres per town
CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
NOX      nitric oxides concentration (parts per 10 million)
RM       average number of rooms per dwelling
AGE      proportion of owner-occupied units built prior to 1940
DIS      weighted distances to five Boston employment centres
RAD      index of accessibility to radial highways
TAX      full-value property-tax rate per $10

In [49]:
X = data.data
features = data.feature_names
y = data.target

Из признаков (характеристик данных) и целевой переменной сформируем датафрейм, в качестве названий колонок возьмем названия признаков.

In [50]:
df = pd.DataFrame(X, columns=features)
df['target'] = y

df.head()

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,target
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222.0,18.7,396.90,5.33,36.2


## Обучение модели

Возьмем линейную модель из `sklearn'a` и обучим на выборке.



In [51]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(X, y)

LinearRegression()

Теперь сделаем предсказания на всей выборке, чтобы понять, насколько хорошо модель научилась предсказывать стоимость дома.

In [54]:
# Преобразуем все признаки в числовой формат
X = X.astype(float)

# Заново обучаем модель и делаем предсказание
lr.fit(X, y)
pred = lr.predict(X)

print(pred[:5]) # Выведем первые 5 предсказаний для проверки

[30.00384338 25.02556238 30.56759672 28.60703649 27.94352423]


## Метрики

Выведем рядом истинные значения целевой переменной и предсказанные моделью значения стоимости дома.

In [55]:
pred_df = pd.DataFrame({
    'ground truth': y,
    'prediction': pred
})

pred_df

,ground truth,prediction
0,24.0,30.003843
1,21.6,25.025562
2,34.7,30.567597
3,33.4,28.607036
4,36.2,27.943524
...,...,...
501,22.4,23.533341
502,20.6,22.375719
503,23.9,27.627426
504,22.0,26.127967


В первой строке должны были предсказать 24.0, а предсказали 30.003843, во второй строке истина 21.6, а модель сказала 25.025562, мы можем пройтись по каждому объекту и посмотреть, насколько близким оказался прогноз.

In [56]:
pred_df['difference'] = pred_df['prediction'] - pred_df['ground truth']
pred_df

,ground truth,prediction,difference
0,24.0,30.003843,6.003843
1,21.6,25.025562,3.425562
2,34.7,30.567597,-4.132403
3,33.4,28.607036,-4.792964
4,36.2,27.943524,-8.256476
...,...,...,...
501,22.4,23.533341,1.133341
502,20.6,22.375719,1.775719
503,23.9,27.627426,3.727426
504,22.0,26.127967,4.127967


Теперь в новом столбце есть разница между предсказанием и истиной. Можем в среднем узнать, насколько эта разница большая: просуммировать отклонения и поделить на количество строк, то есть на 506.

In [57]:
pred_df['difference'].sum() / pred_df.shape[0]

np.float64(1.2357264969740872e-15)

Получилось что-то очень маленькое, максимально приближенное к нулю, значит, что в среднем наши предсказания очень близки к истине, но это не так, ведь выше наблюдали, что предскания отклоняются от истины на приличные значения.  
Усредненные отклонения в среднем дали такую маленькую ошибку из-за того, что разница может быть как положительной (если мы предсказали больше), так и отрицательной (если мы предсказали меньше), а когда мы суммируем положительные и отрицательные значения, они друг друга нивелируют, и по итогу получаем маленькую ошибку.

Получается, что нужно избавиться от знака минуса в наших предсказаниях.

### MAE

 Можем брать абсолютные значения, то есть брать модуль числа:

In [58]:
pred_df['abs'] = abs(pred_df['prediction'] - pred_df['ground truth'])
pred_df

,ground truth,prediction,difference,abs
0,24.0,30.003843,6.003843,6.003843
1,21.6,25.025562,3.425562,3.425562
2,34.7,30.567597,-4.132403,4.132403
3,33.4,28.607036,-4.792964,4.792964
4,36.2,27.943524,-8.256476,8.256476
...,...,...,...,...
501,22.4,23.533341,1.133341,1.133341
502,20.6,22.375719,1.775719,1.775719
503,23.9,27.627426,3.727426,3.727426
504,22.0,26.127967,4.127967,4.127967


Теперь не будет такого, что отрицательные и положительные отклонения схлапываются в ноль.   
Найдем среднее отклонение по модулю для всей выборки:

In [59]:
pred_df['abs'].mean()

np.float64(3.270862810900316)

Получается мы в среднем ошибаемся на 3.27 единиц по всей выборке. Это мы подсчитали метрику под названием MAE (mean absolute error, средне-абсолютная ошибка).

$$MAE = \frac{1}{n}\sum_i^n{|y - y_{pred}|}$$

Она реализована в библиотеки `sklearn` в модуле `metrics`:


In [60]:
from sklearn.metrics import mean_absolute_error

Можем не вручную считать MAE, а с помощью функции `mean_absolute_error`:

In [61]:
mean_absolute_error(pred_df['ground truth'], pred_df['prediction'])

3.270862810900316

Получается такое же значения метрики, как мы получали при ручном подсчете.   
Чем метрика MAE ближе к нулю, тем лучше, в этом случае мы попадаем во все истинные значения.

### MSE

Как ещё мы могли бы избавиться от знака минуса в отклонениях?  
Мы могли бы возводить отклонения в квадрат, а затем так же искать среднее по выборке:

In [62]:
pred_df['squared'] = (pred_df['prediction'] - pred_df['ground truth']) ** 2
pred_df

,ground truth,prediction,difference,abs,squared
0,24.0,30.003843,6.003843,6.003843,36.046135
1,21.6,25.025562,3.425562,3.425562,11.734478
2,34.7,30.567597,-4.132403,4.132403,17.076757
3,33.4,28.607036,-4.792964,4.792964,22.972499
4,36.2,27.943524,-8.256476,8.256476,68.169392
...,...,...,...,...,...
501,22.4,23.533341,1.133341,1.133341,1.284461
502,20.6,22.375719,1.775719,1.775719,3.153178
503,23.9,27.627426,3.727426,3.727426,13.893705
504,22.0,26.127967,4.127967,4.127967,17.040110


Снова не будет отрицательных отклонений, которые ведут к очень маленькому усреднению.   
Найдем среднее отклонение в квадрате для всей выборки:

In [63]:
pred_df['squared'].mean()

np.float64(21.894831181729202)

Получается мы в среднем ошибаемся на 21.89 квадратичных единиц по всей выборке. Эта метрика называется MSE (mean squared error, средне-квадратичная ошибка).

$$MSE = \frac{1}{n}\sum_i^n{(y - y_{pred})^2}$$

Она тоже реализована в библиотеки `sklearn` в модуле `metrics`:


In [64]:
from sklearn.metrics import mean_squared_error

Посчитаем MSE с помощью функции `mean_squared_error`:

In [65]:
mean_squared_error(pred_df['ground truth'], pred_df['prediction'])

21.894831181729202

И снова наши значения совпали.   
Чем метрика MSE ближе к нулю, тем лучше.

Метрика очень классная, она используется под капотом моделей машиного обучения, как функция потерь, но когда хотим понять, насколько модель ошибается, метрика MSE может вносить неясность.

### RMSE

Отклонения считали в квадрате, поэтому если наша стоимость изменяется в рублях, то mse дает цифру рубли в квадрате, что усложняет интерпретацию результата, это легко исправить взятием корня квадратного из значения MSE - а это и есть новая метрика RMSE.

Для подсчета этой метрики можно усреднить квадратичные отклонения (то есть получить MSE), а затем взять корень квадратный:

In [66]:
np.sqrt(pred_df['squared'].mean())

np.float64(4.679191295697281)

Теперь очевидно, что в среднем ошибаемся на 4.67 единиц по всей выборке, либо в большую, либо в меньшую сторону. Эта метрика RMSE (root mean squared error).

$$RMSE = \sqrt{\frac{1}{n}\sum_i^n{(y - y_{pred})^2}}$$

Она не реализована в библиотеки `sklearn`, но как мы поняли, метрика - это MSE  с корнем:


In [67]:
np.sqrt(mean_squared_error(pred_df['ground truth'], pred_df['prediction']))

np.float64(4.679191295697281)

Значения идентичны.
И так же, как предыдущие метрики, чем метрика RMSE ближе к нулю, тем лучше.

### $R^2$

Первые три метрики, которые рассмотрели, имеют идеальное значение 0, к которому нужно стремиться, но эти метрики не имеют пороговых значений, выше которых мы считаем, что модель ужасная, т.к. что они в зависимости от задач могут очень сильно варьироваться.

К примеру, если наша целевая переменная изменяется от 0 до 10:

In [68]:
y_true = [0, 3, 6, 2, 9]

Наша модель может предсказать:

In [69]:
y_pred = [1, 5, 9, 6, 14]

Соединим предсказания и истину вместе и найдем разницу между ними:

In [70]:
df = pd.DataFrame({
    'y_true': y_true,
    'y_pred': y_pred
})

df['difference'] = df['y_pred'] - df['y_true']
df

,y_true,y_pred,difference
0,0,1,1
1,3,5,2
2,6,9,3
3,2,6,4
4,9,14,5


И мы получим следующее значение средне-квадратичной ошибки:

In [71]:
mean_squared_error(y_true, y_pred)

11.0

А если наша целевая переменная изменяется от 0 до 10000:

In [72]:
y_true = [0, 300, 6000, 2000, 9900]

Наша модель может предсказать:

In [73]:
y_pred = [1, 302, 6003, 2004, 9905]

Соединим предсказания и истину вместе и найдем разницу между ними:

In [74]:
df = pd.DataFrame({
    'y_true': y_true,
    'y_pred': y_pred
})

df['difference'] = df['y_pred'] - df['y_true']
df

,y_true,y_pred,difference
0,0,1,1
1,300,302,2
2,6000,6003,3
3,2000,2004,4
4,9900,9905,5


И мы получим следующее значение средне-квадратичной ошибки:

In [75]:
mean_squared_error(y_true, y_pred)

11.0

MSE получились одинаковыми, потому что одинаковые разницы в предсказаниях, но вторая модель очевидно делает меньше ошибок, чем первая, значит она более качественная, хотя значения метрики MSE абсолютно одинаковые.

MSE не позволяет сама по себе сделать вывод о том, как хорошо модель решает задачу. Вот как в нашем примере, если целевая переменная принимает значения от 0 до 10, а MSE равняется 11, это плохой показатель, а когда целевая переменная варьируется от 0 до 10000, такое же значение уже является очень хорошим.  
Для избавления от такой неясности и был введен коэффициент детерминации, который по сути является нормированной среднеквадратичной ошибкой и принимает значения от 0 до 1.


$$R^2 = 1 - \frac{\frac{1}{n}\sum^{n}_{i}{(y - y_{pred})^2}}{\frac{1}{n}\sum^{n}_{i}{(y - \bar{y})^2}}$$

где $\bar{y}=\frac{1}{l}\sum^{n}_{i}y_{i}$ - среднее значение целевой переменной.

Коэффициент детерминации характеризует, какую долю дисперсии ответов объясняет модель. Если $R^{2}=1$, то модель идеально описывает данные, если же $R^{2}$ близко к нулю, то предсказания сопоставимы по качеству с константной моделью.  
То есть, можем относиться к метрике $R^{2}$ как к точности работы модели для задачи регрессии.

Для подсчета этой метрики в числителе используется MSE нашей модели, а в знаменателе дисперсия целевой переменной (хотя к этому значению можно относиться, как к MSE не самой умной константной модели, которая везде отвечает среднее по выборке):

In [76]:
pred_df['constant'] = pred_df['ground truth'].mean()
pred_df

,ground truth,prediction,difference,abs,squared,constant
0,24.0,30.003843,6.003843,6.003843,36.046135,22.532806
1,21.6,25.025562,3.425562,3.425562,11.734478,22.532806
2,34.7,30.567597,-4.132403,4.132403,17.076757,22.532806
3,33.4,28.607036,-4.792964,4.792964,22.972499,22.532806
4,36.2,27.943524,-8.256476,8.256476,68.169392,22.532806
...,...,...,...,...,...,...
501,22.4,23.533341,1.133341,1.133341,1.284461,22.532806
502,20.6,22.375719,1.775719,1.775719,3.153178,22.532806
503,23.9,27.627426,3.727426,3.727426,13.893705,22.532806
504,22.0,26.127967,4.127967,4.127967,17.040110,22.532806


Получаем, что MSE константной модели, которая везде отвечает среднее по выборке равно

In [77]:
mse_const = mean_squared_error(pred_df['ground truth'], pred_df['constant'])
mse_const

84.41955615616556

А MSE нашей модели получается следующим

In [78]:
mse_model = mean_squared_error(pred_df['ground truth'], pred_df['prediction'])
mse_model

21.894831181729202

Что лучше, чем у константных ответов, значит, наша модели описывает дисперсию целевой переменной, значит, что она уловила больше вариативности в данных, нежели простое среднее по выборке.   
Теперь можем посчитать коэффициент детерминации: поделим MSE нашей модели на дисперсию целевой переменной, а затем это значение вычтем из 1:

In [79]:
1 - mse_model / mse_const

0.7406426641094095

$R^2$ реализована в библиотеке `sklearn`:

In [80]:
from sklearn.metrics import r2_score

r2_score(pred_df['ground truth'], pred_df['prediction'])

0.7406426641094095

Получается $R^2$ нашей модели стремится к 1, что говорит о довольно неплохом качестве работы модели.

#### Отрицательный $R^2$

Иногда случается так, что модель получается очень плохой, что её ответы никак не соотносятся с истинными, в этом случае коэффициент детерминации может быть отрицательным.


In [81]:
y_true = [0, 1, 2, 3, 4]
y_pred = [10, -20, -30, 40, -50]


df = pd.DataFrame({
    'y_true': y_true,
    'y_pred': y_pred
})


df['difference'] = df['y_pred'] - df['y_true']
df

,y_true,y_pred,difference
0,0,10,10
1,1,-20,-21
2,2,-30,-32
3,3,40,37
4,4,-50,-54


In [82]:
variance = ((df['y_true'] - df['y_true'].mean()) ** 2).mean()
print(f'Дисперсия выборки {variance}')

Дисперсия выборки 2.0


In [83]:
mse = ((df['y_true'] - df['y_pred']) ** 2).mean()
print(f'MSE {mse}')

MSE 1170.0


Полученная MSE гораздо больше дисперсии выборки, из-за этого и получается большое число при делении MSE на MSE константной модели, которая везде отвечает среднее по выборке.

In [84]:
mse / variance

np.float64(585.0)

Вычитая полученное число из 1 и получается отрицательное значение:

In [85]:
1 - mse / variance

np.float64(-584.0)

Или же сразу могли посчитать метрику $R^2$ через sklearn.

In [86]:
r2_score(y_true, y_pred)

-584.0

Пугаться этого не стоит, а нужно исправить модель, возможно модель переобучилась, либо при предсказывании объектов была использована неправильная предобработка данных.

## Summary

<table>

<tr>
<td>
Метрика
</td>

<td>
Формула
</td>

<td>
Диапазоны значений
</td>

<td>
Идеальное значение
</td>
</tr>

<tr>
<td>
MAE (mean absolute error, средне-абсолютная ошибка)
</td>

<td>
$$MAE = \frac{1}{n}\sum_i^n{|y - y_{pred}|}$$

</td>

<td>
[0, +$\infty$)
</td>

<td>
0
</td>
</tr>

<tr>
<td>
MSE (mean squared error, средне-квадратичная ошибка)

</td>

<td>
$$MSE = \frac{1}{n}\sum_i^n{(y - y_{pred})^2}$$

</td>

<td>
[0, +$\infty$)
</td>

<td>
0
</td>
</tr>

<tr>
<td>
RMSE (root mean squared error).

</td>

<td>
$$RMSE = \sqrt{\frac{1}{n}\sum_i^n{(y - y_{pred})^2}}$$

</td>

<td>
[0, +$\infty$)
</td>

<td>
0
</td>
</tr>

<tr>
<td>
Коэффициент детерминации $R^{2}$

</td>

<td>
$$R^2 = 1 - \frac{\frac{1}{n}\sum^{n}_{i}{(y - y_{pred})^2}}{\frac{1}{n}\sum^{n}_{i}{(y - \bar{y})^2}}$$

</td>

<td>
(-$\infty$, 1]

</td>

<td>
1
</td>
</tr>
</table>

Метрики используются для того, чтобы получить цифру оценки работы обученной модели, чтобы уже заранее знать, чего ожидать от модели, когда ей будем давать реальные данные, а так же метрики мы можем использовать для обнаружения переобучения/недообучения. Про них разговор ведется в этом [ноутбуке](https://colab.research.google.com/drive/1MNpLMZGJCBwdNjY_PgJIKHtll6zvUEyU?usp=sharing) и в этом видео.

Муррр ♥